From a222c818c7d3daddac9acab27d9983d46fd29708 Mon Sep 17 00:00:00 2001
From: James Le Cuirot <chewi@gentoo.org>
Date: Wed, 27 May 2026 11:44:22 +0100
Subject: [PATCH] Add FEATURES="buildpkg-proactive" to rebuild missing/unusable
 binpkgs

If you build binary packages in order to do a --usepkgonly build later,
you want to be sure that no packages are missing and that the packages
you do have will actually be visible, e.g. they were not built when a
now stable package was previously unstable.

This will proactively rebuild matching packages that have no (usable)
binary package, even if those packages are already installed.

This required some refactoring of the logic around whether to build a
binary package or not. Note that it is not necessary to check whether
--buildpkg=n was given because adjust_config() already takes care of
that.

I thought about automatically enabling buildpkg-proactive when
--buildpkgonly is given, but that option applies to / as well as ROOT,
so this would have been unhelpful.

Bug: https://bugs.gentoo.org/976083
Signed-off-by: James Le Cuirot <chewi@gentoo.org>
--- a/lib/_emerge/EbuildBuild.py
+++ b/lib/_emerge/EbuildBuild.py
@@ -327,40 +327,15 @@ class EbuildBuild(CompositeTask):
         self._build_dir.clean_log()
         pkg = self.pkg
         logger = self.logger
-        opts = self.opts
         pkg_count = self.pkg_count
         scheduler = self.scheduler
         settings = self.settings
-        features = settings.features
         ebuild_path = self._ebuild_path
-        system_set = pkg.root_config.sets["system"]
 
         # buildsyspkg: Check if we need to _force_ binary package creation
-        self._issyspkg = (
-            "buildsyspkg" in features
-            and system_set.findAtomForPackage(pkg)
-            and "buildpkg" not in features
-            and opts.buildpkg != "n"
-        )
-
-        # Do not build binary cache for packages from volatile sources.
-        # For volatile sources (eg., git), the PROPERTIES parameter in
-        # the ebuild is set to 'live'.
-
-        # The default behavior is to build binary cache for all pkgs.
-        # "buildpkg-live" is a FEATURE that is enabled by default.
-        # To not build binary cache for live pkgs, we disable it by
-        # specifying FEATURES="-buildpkg-live"
+        self._issyspkg = pkg.syspkg_wanted()
 
-        buildpkg_live = "buildpkg-live" in features
-        live_ebuild = "live" in self.settings.get("PROPERTIES", "").split()
-        buildpkg_live_disabled = live_ebuild and not buildpkg_live
-
-        if (
-            ("buildpkg" in features or self._issyspkg)
-            and not buildpkg_live_disabled
-            and not self.opts.buildpkg_exclude.findAtomForPackage(pkg)
-        ):
+        if pkg.binpkg_wanted(self.opts.buildpkg_exclude):
             self._buildpkg = True
 
             msg = " === ({} of {}) Compiling/Packaging ({}::{})".format(
--- a/lib/_emerge/Package.py
+++ b/lib/_emerge/Package.py
@@ -601,6 +601,34 @@ class Package(Task):
         s += ")"
         return s
 
+    def syspkg_wanted(self):
+        """Tests whether this is a system package to build a binary package for
+        (buildsyspkg), assuming it isn't already covered by buildpkg."""
+        features = self.root_config.settings.features
+        return (
+            "buildsyspkg" in features
+            and "buildpkg" not in features
+            and self.root_config.sets["system"].findAtomForPackage(self)
+        )
+
+    def binpkg_wanted(self, exclude):
+        """Tests whether this is a package to build a binary package for, taking
+        account of FEATURES, PROPERTIES, and the given excluded set."""
+        # Do not build binary cache for packages from volatile sources.
+        # For volatile sources (eg., git), the PROPERTIES parameter in
+        # the ebuild is set to 'live'.
+        #
+        # The default behavior is to build binary cache for all pkgs.
+        # "buildpkg-live" is a FEATURE that is enabled by default.
+        # To not build binary cache for live pkgs, we disable it by
+        # specifying FEATURES="-buildpkg-live"
+        features = self.root_config.settings.features
+        return (
+            ("buildpkg-live" in features or not "live" in self.properties)
+            and not exclude.findAtomForPackage(self)
+            and ("buildpkg" in features or self.syspkg_wanted())
+        )
+
     class _use_class:
         __slots__ = ("enabled", "_expand", "_expand_hidden", "_force", "_pkg", "_mask")
 
--- a/lib/_emerge/depgraph.py
+++ b/lib/_emerge/depgraph.py
@@ -216,6 +216,12 @@ class _frozen_depgraph_config:
         atoms = " ".join(myopts.get("--rebuild-ignore", [])).split()
         self.rebuild_ignore = _wildcard_set(atoms)
 
+        self.buildpkg_exclude = InternalPackageSet(
+            initial_atoms=" ".join(myopts.get("--buildpkg-exclude", [])).split(),
+            allow_wildcard=True,
+            allow_repo=True,
+        )
+
         self.rebuild_if_new_rev = "--rebuild-if-new-rev" in myopts
         self.rebuild_if_new_ver = "--rebuild-if-new-ver" in myopts
         self.rebuild_if_unbuilt = "--rebuild-if-unbuilt" in myopts
@@ -7804,6 +7810,22 @@ class depgraph:
                                 pkg, autounmask_level=autounmask_level
                             ):
                                 continue
+                            # Also reject built instances if we want a binary
+                            # package, but none exist or all the existing ones
+                            # are now masked, e.g. accepted keywords changed.
+                            elif (
+                                "buildpkg-proactive" in root_config.settings.features
+                                and pkg.binpkg_wanted(
+                                    self._frozen_config.buildpkg_exclude
+                                )
+                            ):
+                                for binpkg in self._iter_match_pkgs_atom(
+                                    root_config, "binary", Atom(f"={pkg.cpv}")
+                                ):
+                                    if not binpkg.masks:
+                                        break
+                                else:
+                                    continue
 
                     # Calculation of USE for unbuilt ebuilds is relatively
                     # expensive, so it is only performed lazily, after the
--- a/lib/portage/const.py
+++ b/lib/portage/const.py
@@ -167,6 +167,7 @@ SUPPORTED_FEATURES = frozenset(
         "binpkg-signing",
         "buildpkg",
         "buildpkg-live",
+        "buildpkg-proactive",
         "buildsyspkg",
         "candy",
         "case-insensitive-fs",
--- a/man/make.conf.5
+++ b/man/make.conf.5
@@ -414,6 +414,11 @@ When this option is enabled (the default), \fBbuildpkg\fR will exhibit
 the default behavior of building binary cache for all packages. When
 it is disabled, binary packages will not be created for live ebuilds.
 .TP
+.B buildpkg\-proactive
+Proactively rebuild matching packages that have no (usable) binary package,
+even if those packages are already installed. This does nothing unless
+\fBbuildpkg\fR or \fBbuildsyspkg\fR is also enabled.
+.TP
 .B buildsyspkg
 Build binary packages for just packages in the system set.
 .TP
-- 
2.54.0

